### PR TITLE
Auto timestamp existing log files instead of overwriting

### PR DESCRIPTION
**Pull request by 0Nel**
Related to https://github.com/rock-core/tools-logger/pull/8

Instead of incrementing the existing logfile a timestamp suffix is added to the existing logfile (if additional auto_rename_existing_file property is true).

In contrast to incrementing, timestamping is more practical in quickly identify recorded log files. Furthermore, especially for a large numbers of recorded files, it is quicker to add a unique timestamp instead of tediously increment through all existing files. The implemented approach assumes that not more than one log file with the same basic _file property is started per second. To be certain, the existence of the timestamped file is checked again and returns false in the startHook if existing.

I adapted the test suite to consider the new task property and handle a possible change of the _file property. To keep the Pocolog stream working without changing the general setup of the test_suite I had to change from Tempfile.open to File.open to prevent Tempfile from adding its own unique tag to the file name. Alternatively one could set @logfile_io.path directly. This is more unstable as it requires that logfile_path() was previously called.

I also adapted the test suite to consider the rock_cxx_type_name to respect the change of metadata introduced in https://github.com/rock-core/tools-logger/commit/1067db83174d5f33a417aaed3800d8ca693b1080

### CODE REVIEW

File: test/test_logger.rb [hunk 1]
```diff
@@ -27,15 +28,22 @@ def teardown
...
     end
 
-    def logfile_path
-        @logfile_io ||= Tempfile.open('rock_logger_test.0.log')
+    def logfile_path(arg='/tmp/rock_logger_test.log')
+        @logfile_io ||= File.open(arg, "w+")
+        unless @logfile_io.path == arg
+          @logfile_io.close
+          @logfile_io = File.open(arg, "w+")
+        end
```
Comment (doudou) [inline test/test_logger.rb]: Indentation in this file is 4 spaces.

File: tasks/Logger.cpp [hunk 2]
```diff
@@ -56,13 +61,38 @@ Logger::~Logger()
...
+        vector<string> strs;
+        boost::split(strs, _file.value(), boost::is_any_of("."));
+        strs.insert(strs.end()-1, std::string(suffix));
+        // safety check if timestamped file exists
+        std::string timestamped_str = boost::algorithm::join(strs, ".");
+        if(boost::filesystem::exists(timestamped_str)) {
+          log(Error) << "Timestamped file " << _file.value() << " already exists. Please retry." << endlog();
+          return false;
+        }
+        _file.set(timestamped_str);
```
Comment (doudou) [inline tasks/Logger.cpp]: Dynamically changing a property from within the task is really not an expected behavior.

You could create an `attribute` for this (`current_file`). Same general API, just replacing `property` by `attribute` in the orogen file.

File: tasks/Logger.cpp [hunk 3]
```diff
@@ -56,13 +61,38 @@ Logger::~Logger()
...
+        log(Warning) << "Writing to " << _file.value() << " ." << endlog();
```
Comment (doudou) [inline tasks/Logger.cpp]: Why would this be a warning ? Isn't this Info at best ?

File: test/test_logger.rb [hunk 4]
```diff
@@ -14,7 +14,8 @@ def task
 
     def setup
         super
-        task.overwrite_existing_files = true
+        task.overwrite_existing_files = false
+        task.auto_rename_existing_files = true
```
Comment (doudou) [inline test/test_logger.rb]: Unless I'm missing something, this is the _only_ use case you are testing now.

Make a common behavior case and subclass it for the two possible configurations (three if you're feeling like writing tests for overwrite_existing_files).

File: test/test_logger.rb [hunk 5]
```diff
@@ -27,15 +28,22 @@ def teardown
...
+          @logfile_io.close
+          @logfile_io = File.open(arg, "w+")
+        end
         @logfile_io.path
     end
 
-    def logfile
-        @logfile ||= Pocolog::Logfiles.open(logfile_path)
+    def logfile(file)
+        @logfile ||= Pocolog::Logfiles.open(logfile_path(file))
```
Comment (doudou) [inline test/test_logger.rb]: This whole behind-the-scenes magic makes things **really** hard to follow and understand. Now that we can't assume that `.file` is a constant (or rather `.current_file` ;-)), change the test suite accordingly: open the log file from within the test suite at the point of use, without storing it at all as an attribute. This will remove the need to check for the `auto_*` attribute in all test cases, and make the subclass-and-set-auto-property-to-true a clean thing.

You're actually already mostly doing that when you do logfile(path).stream() everywhere.

**Comment by 0Nel** [comment]
Hi Sylvain,

thank you for your comments and the thought food ;)

I am going to update the code accordingly. I have a few further thoughts and comments tho.

    ...the general scheme followed by e.g. Syskit is to create one timestamped folder
       with all the data of the given run, instead of timestamping every file.

Similar to the syskit scheme, this issue is already well cared for, when using Bundles in logging scripts, which automatically creates timestamped folders. However the aimed use case for this implementation is related to repeated, manual data logging, e.g. sequential logging through rock-display. When logging in this manner it is actually really beneficial to later match logs with experiments using timestamps. As creating directories from within the task is not really an option I would stick to timestamping files, which is handled likewise in rosbags.

      Another thing about the filename.X.log scheme is that pocolog and other Rock tools
      support this scheme out of the box (it will allow loading them as a sequence).

I was not aware of this. Whats your opinion on prepending the timestamp? Might be confusing, but seems to be the most stable solution, since injecting the timestamp before the index, would mean to assume that more than one point delimiter is present within the file string.

      As for the seconds resolution issue, I think it might introduce spurious errors.

Option 1: Return to minute resolution and increment integer until filename is available. Clean and stable, but potentially slow.
Option 2: Add a unique hash, test for file existence. Not safe, but quick.
Option 3: Live with the current solution as it improves the aimed manual use case and stick to scripted logging and use of Bundles to generate clean repeated logs. Worst case is the Task returning an Error about the already existing timestamped file, which is the default behavior of the current state.

If I want to move the file renaming to a, e.g. get_available_filename, method. Should I use an operation for this? Not really familiar with rocks best practice yet.

File: tasks/Logger.cpp (hunk 2 above)
Comment (0Nel) [reply tasks/Logger.cpp]: Ok

File: tasks/Logger.cpp (hunk 3 above)
Comment (0Nel) [reply tasks/Logger.cpp]: We can change to Info. My original thought was that changing the property from within the code may be confusing and the user should be warned about the automatic renaming. The automatic_renaming_* property makes the warning dispensable, but I would say it is worth an Info to notify about the final log file name.

File: test/test_logger.rb (hunk 4 above)
Comment (0Nel) [reply test/test_logger.rb]: True. I was manually changing these values for debugging. I am gonna stick to default values during setup and write specific tests.

File: test/test_logger.rb (hunk 5 above)
Comment (0Nel) [reply test/test_logger.rb]: Agreed. I was quit unhappy with passing around arguments. Will strip this down to logfile(path).stream().

**Comment by doudou** [comment]
Option 4: check if the file exists, and if it does wait one second and retry
Option 5: move the renaming to a post-processing step (opens file, check time of first sample and renames the file accordingly)

>  If I want to move the file renaming to a, e.g. get_available_filename, method. Should I use an operation for this? Not really familiar with rocks best practice yet.

Not sure what you mean by that. If it's meant to be used internally, just a standard C++ method will do just fine. If not, I'm not sure what's the use-case.

**Comment by doudou** [comment]
Ping @0Nel. Just to clarify, the options 4 and 5 were meant as suggestions. I think (4) is a good middle-ground, as in principle it should happen rarely.

**Comment by 0Nel** [comment]
Updated the task and test with respect to our discussion.

Whats your opinion on this matter?

 > ```
>   Another thing about the filename.X.log scheme is that pocolog and other Rock tools
>   support this scheme out of the box (it will allow loading them as a sequence).
> ```
> I was not aware of this. Whats your opinion on prepending the timestamp? Might be confusing, but seems to be the most stable solution, since injecting the timestamp before the index, would mean to assume that more than one point delimiter is present within the file string.

**Comment by 0Nel** [comment]
I had to fix my previously unfunctional implementation of Option 4 (retimestamping).

Still not quite happy with falling back to the file property in the scope of the renameFile() method, but slicing and replacing the old timestamp seems to be too much effort, as long as the file property is still available.

Also added another test for this to the test suite.

File: test/test_logger.rb [hunk 6]
```diff
@@ -3,6 +3,7 @@
 
 require 'pocolog'
 require 'fileutils'
+require 'pry'
```
Comment (doudou) [inline test/test_logger.rb]: That's debugging stuff, don't keep it there.

File: test/test_logger.rb [hunk 7]
```diff
@@ -80,6 +72,36 @@ def test_basics
...
+    end
+
+    def test_ambiguous_properties
+        task.overwrite_existing_files = true
+        task.auto_rename_existing_files = true
+        task.configure
+        assert_raises Orocos::StateTransitionFailed do
+            task.start
+        end
+        puts '[INFO] This Error is expected and handled.'
```
Comment (doudou) [inline test/test_logger.rb]: Don't puts in tests. If you want something informative, make the test name more explicit and run with `-v` to see the test progress.

File: test/test_logger.rb [hunk 8]
```diff
@@ -80,6 +72,36 @@ def test_basics
...
 
+    def test_no_overwrite
+        task.overwrite_existing_files = false
+        task.auto_rename_existing_files = false
+        task.configure
+        touch_file = File.new(task.file, "w")
+        assert_raises Orocos::StateTransitionFailed do
+            task.start
+        end
+        puts '[INFO] This Error is expected and handled.'
```
Comment (doudou) [inline test/test_logger.rb]: Don't puts in tests. If you want something informative, make the test name more explicit and run with -v to see the test progress.

File: tasks/Logger.cpp [hunk 9]
```diff
@@ -67,37 +67,45 @@ bool Logger::startHook()
...
     }
 
-    if(boost::filesystem::exists(_file.value()) && !_overwrite_existing_files.get() && _auto_rename_existing_files.get())
+    if(boost::filesystem::exists(_current_file.get()) && !_overwrite_existing_files.get() && !_auto_rename_existing_files.get())
+    {
+      log(Error) << "File " << _current_file.get() << " already exists. No overwrite allowed by task property." << endlog();
+      return false;
+    }
+
+    if(boost::filesystem::exists(_current_file.get()) && !_overwrite_existing_files.get() && _auto_rename_existing_files.get())
```
Comment (doudou) [inline tasks/Logger.cpp]: This actually makes the overall behavior inconsistent. The first file will never have a timestamp, only the following ones.

I'd suggest to renaming the property to _auto_timestamp_files and always add a timestamp when it is set.

**Comment by 0Nel** [comment]
> Almost there.

I guess we're there now :wink: 

PS: The logger doesn't have it's own `configureHook`, but I think we meant the same place.

**Review by doudou — APPROVED** [review]

**Comment by doudou** [comment]
Given how critical this is, I'd appreciate a final review @2maz ...

File: tasks/Logger.cpp [hunk 10]
```diff
@@ -349,3 +367,47 @@ void Logger::snapshot()
...
+{
+    // create timestamp
+    time_t now = time(0);
+    tm *t_ptr = localtime(&now);
+    char suffix[21];
+    strftime(suffix, sizeof(suffix), "%F_%H-%M-%S", t_ptr);
+    // append suffix to previous _current_file.get()
+    vector<string> strs;
+    boost::split(strs, _file.value(), boost::is_any_of("."));
+    strs.insert(strs.end()-1, std::string(suffix));
```
[the remark this answers is not in the document]
Comment (0Nel) [reply tasks/Logger.cpp]: Good call! For files without a suffix the timestamp would have been prepended. Fixed this in new commit.

File: tasks/Logger.cpp [hunk 11]
```diff
@@ -349,3 +367,50 @@ void Logger::snapshot()
...
+    char suffix[21];
+    strftime(suffix, sizeof(suffix), "%F_%H-%M-%S", t_ptr);
+    // append suffix to previous _current_file.get()
+    vector<string> strs;
+    boost::split(strs, _file.value(), boost::is_any_of("."));
+    if ( strs.size() == 1) {
+        strs.insert(strs.end(), std::string(suffix));
+    } else {
+        strs.insert(strs.end()-1, std::string(suffix));
+    }
```
Comment (doudou) [inline tasks/Logger.cpp:384]: Would be good to add a test for this one ...

Comment (0Nel) [reply tasks/Logger.cpp:384]: Done :+1:

**Comment by 0Nel** [comment]
I just stumbled over the auto_timestamp_files property again. At the moment it is only evaluated, if the currently defined file already exists. I can see how this might end up being confusing. Should I rewrite it in a way that any file would be timestamped as long as the property is set to true, even if the original file does not yet exist?

File: test/test_logger.rb [hunk 12]
```diff
@@ -113,6 +113,17 @@ def test_re_stamping_existing_timestamped_file
...
 
+    def test_no_suffix_log_file
+        task.overwrite_existing_files = false
+        task.auto_timestamp_files = true
+        task.file = "/tmp/rock_logger_test"
+        touch_file = File.new(task.file, "w")
+        assert(!task.has_port?('time'))
+        assert(task.createLoggingPort('time', '/base/Time', []))
+        generate_and_check_logfile
+        assert(task.file != task.current_file)
```
Comment (doudou) [inline test/test_logger.rb:124]: You're actually not testing the bug that has been fixed, that is that the timestamp will be added as a suffix.

**Review by doudou — APPROVED** [review]

File: test/test_logger.rb (hunk 12 above)
Comment (0Nel) [reply test/test_logger.rb]: My bad. Should be resolved now..

**Comment by doudou** [comment]
> Should I rewrite it in a way that any file would be timestamped as long as the property is set to true, even if the original file does not yet exist?

Good catch. I think so, yes.

**Comment by doudou** [comment]
I'm still good with it. @2maz ?

File: tasks/Logger.cpp [hunk 13]
```diff
@@ -349,3 +372,47 @@ void Logger::snapshot()
...
+{
+    // create timestamp
+    time_t now = time(0);
+    tm *t_ptr = localtime(&now);
+    char suffix[21];
+    strftime(suffix, sizeof(suffix), "%F_%H-%M-%S", t_ptr);
+    // append suffix to previous _file.value()
+    vector<string> strs;
+    boost::split(strs, _file.value(), boost::is_any_of("."));
+    if ( strs.size() == 1) {
```
Comment (2maz) [inline tasks/Logger.cpp:385]: inconsistent bracket style

**Review by 2maz — APPROVED** [review]
Minor style issue remaining, but otherwise fine

**Review by doudou — APPROVED** [review]

**Comment by doudou** [comment]
Thanks for having the patience to go through such a lengthy review process, @0Nel

**Comment by 0Nel** [comment]
You're welcome. Happy to see that we finally came to an end :)
